astropy/units/format/base.py

```diff
@@ -52,7 +52,7 @@ def _fraction_formatters(cls) -> dict[bool | str, Callable[[str, str, str], str]
 
     @classmethod
     def format_exponential_notation(
-        cls, val: float | np.number, format_spec: str = ".8g"
+        cls, val: UnitScale | np.number, format_spec: str = ".8g"
```

Comment:
This method has a docstring, but that only documents the type of `val` very vaguely as `"number"`. When I added the annotation I inserted `isinstance()` checks to this method and ran the tests to see what types were actually encountered at runtime. It turns out that the annotation was overly restrictive, but the fact that the runtime checks I made didn't catch this means that we are missing tests for string conversion of units with scale factors that are `fractions.Fraction` or `complex` instances. I will leave it to the subpackage maintainers to decide if such tests should be added.

Reply:
Given that this is a private function, I don't think the documentation needs to be changed, but tests would be a useful addition.

Reply:
Hmm, that's weird - a complex scale is indeed possible (e.g., `(u.mag()**0.5).decompose()`), but a Fraction is surprising.

Note that with a large complex number, this code does do the wrong thing:
```
u.format.base.Base.format_exponential_notation(1e20+1e20j)
# '1'
```
Something must have gone wrong in some of the refactoring, since this is not a problem in 6.1: 
```
# '1e+20+1e+20j'
```
Fortunately, generally, one does not hit this because `Generic` overrides the method.

Overall, this suggests some tests would be appropriate... Indeed, right now there are problems:
```
s = u.CompositeUnit(1e20+1e20j, [], [])
s.to_string()
'1e+20+1e+20j'
s.to_string('cds')
''  # ????? Likely used the above and replaced 1 with nothing.
s.to_string('fits')
---> 80 if base % 1.0 != 0.0:
TypeError: unsupported operand type(s) for %: 'numpy.complex128' and 'float'
s.to_string('ogip')
--> 352     if math.log10(unit.scale) % 1.0 != 0.0:
TypeError: must be real number, not complex
s.to_string('vounit')
UnitScaleError: The VOUnit format is not able to represent scale for dimensionless units. Multiply your data by 1.000000e+20+1.000000e+20j.
```

These are regressions: In 6.1, 'ogip', 'fits', 'cds', and 'vounit' all gave,
```
TypeError: (1e+20+1e+20j) can not be converted to a Unit
```
which is a much better exception.

That said, I'm not sure it would be wise to address that in this PR...

Reply:
> Hmm, that's weird - a complex scale is indeed possible (e.g., `(u.mag()**0.5).decompose()`), but a Fraction is surprising.

I don't know when or why it was decided that `Fraction` should be an allowed type for scale factors, but I can point out a comment you yourself have written: https://github.com/astropy/astropy/blob/7896e82c0b06ea4b3870711e7124569f812df39f/astropy/units/utils.py#L213


> Note that with a large complex number, this code does do the wrong thing:
> 
> ```
> u.format.base.Base.format_exponential_notation(1e20+1e20j)
> # '1'
> ```
> 
> Something must have gone wrong in some of the refactoring, since this is not a problem in 6.1:
> 
> ```
> # '1e+20+1e+20j'
> ```

Changing the behavior of `Base.format_exonential_notation()` was a deliberate decision in 1895d9c430ef01fb8166ceb876b0aca35ff29805. I see no evidence of a refactoring mistake here.

> Overall, this suggests some tests would be appropriate... Indeed, right now there are problems:
> 
> ```
> s = u.CompositeUnit(1e20+1e20j, [], [])
> s.to_string()
> '1e+20+1e+20j'
> s.to_string('cds')
> ''  # ????? Likely used the above and replaced 1 with nothing.
> s.to_string('fits')
> ---> 80 if base % 1.0 != 0.0:
> TypeError: unsupported operand type(s) for %: 'numpy.complex128' and 'float'
> s.to_string('ogip')
> --> 352     if math.log10(unit.scale) % 1.0 != 0.0:
> TypeError: must be real number, not complex
> s.to_string('vounit')
> UnitScaleError: The VOUnit format is not able to represent scale for dimensionless units. Multiply your data by 1.000000e+20+1.000000e+20j.
> ```
> 
> These are regressions: In 6.1, 'ogip', 'fits', 'cds', and 'vounit' all gave,
> 
> ```
> TypeError: (1e+20+1e+20j) can not be converted to a Unit
> ```
> 
> which is a much better exception.

The commit that changed the behavior is 8cdb9eb26ebf7ca7cdf1f553a3d75c62f0f00370. I would say this reveals a bug in `u.Unit()`, which is that it is too restrictive about the types that are allowed for the creation of a dimensionless unit with a scale factor: https://github.com/astropy/astropy/blob/7896e82c0b06ea4b3870711e7124569f812df39f/astropy/units/core.py#L2135-L2136